File: ezbroWorld/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        # when n = [1,2,3,4],
        # output = [n[1]*n[2]*n[3], n[0] * n[2]*n[3], n[0]*n[1] * n[3], n[0]*n[1]*n[2]]
        result = [1 for i in range(len(nums))]
        leftToRight = [1 for i in range(len(nums))]
        rightToLeft = [1 for i in range(len(nums))]
        tmp1,tmp2 = nums[0], nums[len(nums)-1]
        
        # 1, 2, 3
        for i in range(1,len(nums)):
            leftToRight[i] *= tmp1
            tmp1 *= nums[i]
            
        # 2, 1, 0
        for i in range(len(nums)-2,-1,-1):
            rightToLeft[i] *= tmp2
            tmp2 *= nums[i]
            
        for i in range(len(nums)):
            result[i] = leftToRight[i]*rightToLeft[i]
            
        return result

File: ezbroWorld/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] leftToRight = new int[length];
        int[] rightToLeft = new int[length];
        int[] outputs = new int[length];

        Arrays.fill(leftToRight,1);
        Arrays.fill(rightToLeft,1);
        
        for(int i = 1; i<length;i++){
            leftToRight[i] = leftToRight[i-1] * nums[i-1];
        }
        
        for(int j = length-2 ; j>-1;j--){
            rightToLeft[j] = rightToLeft[j+1] * nums[j+1];
        }
        
        for(int i =0; i<length;i++){
            outputs[i] = leftToRight[i] * rightToLeft[i];
        }
                
        return outputs;
    }
}
